iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp

```diff
@@ -0,0 +1,98 @@
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.h"
+
+#include "emitc/Dialect/EmitC/EmitCDialect.h"
+#include "iree/compiler/Dialect/VM/IR/VMOps.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/Module.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+// Taken over from StandartToVM.
+// We need to replace the Op depending on the operand.
+// We could start with a conversion for IREE::VM::AddI32Op
+template <typename SrcOpTy, typename DstOpTy>
+class BinaryArithmeticOpConversion : public OpConversionPattern<SrcOpTy> {
+  using OpConversionPattern<SrcOpTy>::OpConversionPattern;
+
+ public:
+  BinaryArithmeticOpConversion(MLIRContext *context, StringRef funcName)
+      : OpConversionPattern<SrcOpTy>(context), funcName(funcName) {}
+
+ private:
+  LogicalResult matchAndRewrite(
+      SrcOpTy srcOp, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    typename SrcOpTy::Adaptor srcAdapter(operands);
+
+    MLIRContext *ctx = srcOp.getContext();
+
+    // name of the function to call
+    StringAttr callee_ = StringAttr::get(funcName, ctx);
+
+    // attributes of the function call; references only the variable operands
+    ArrayAttr args_ = ArrayAttr::get({IntegerAttr::get(IndexType::get(ctx), 0),
+                                      IntegerAttr::get(IndexType::get(ctx), 1)},
+                                     ctx);
+    // operands of the function
+    ValueRange operands_{srcAdapter.lhs(), srcAdapter.rhs()};
+
+    rewriter.replaceOpWithNewOp<DstOpTy>(srcOp, srcAdapter.lhs().getType(),
+                                         callee_, args_, operands_);
+
+    return success();
+  }
+
+  StringRef funcName;
+};
+
+}  // namespace
+
+void populateVMToCPatterns(MLIRContext *context,
+                           OwningRewritePatternList &patterns) {
+  patterns.insert<
+      BinaryArithmeticOpConversion<IREE::VM::AddI32Op, mlir::emitc::CallOp>>(
+      context, "vm_add_i32");
+}
+
+namespace IREE {
+namespace VM {
+
+// A pass converting IREE VM operations into the EmitC dialect.
+void ConvertVMToEmitCPass::runOnOperation() {
+  ConversionTarget target(getContext());
+
+  OwningRewritePatternList patterns;
+  populateVMToCPatterns(&getContext(), patterns);
+
+  target.addLegalDialect<mlir::emitc::EmitCDialect>();
+  target.addLegalDialect<IREE::VM::VMDialect>();
+  target.addIllegalOp<IREE::VM::AddI32Op>();
+
+  if (failed(applyPartialConversion(getOperation(), target, patterns))) {
```

Comment:
Should this be a full conversion? They tend to give nicer errors, so if we expect everything to end up in the legal set, I'd recommend using that. It looks like you've already done the work of defining the VMDialect as legal (which I assume will go away once all the conversions are in place), so you should be able to make this a full conversion and it will just legalize the add op.

Reply:
Thanks, for the advice :) You're right, as soon as all conversions are in place, it was planned to change this to a full conversion. If that's already possible because the VM dialect is defined as legal, we should already do this. @simon-camp Do you want to take a look?

Reply:
Yeah I switched to a full conversion and also changed the Pass to run on an IREE Module instead of a std Module.

@GMNGeoffrey Just for clarification.

Reply:
You can modify the operation itself in an operation pass, but you can't replace it because that would be adding a new operation to the parent region. The pass as-is is fine. If you want a rewrite of an IREE module to happen, I think you would need the pass to operate on whatever the parent of that is (probably a std Module).